### PR TITLE
ddl: Add table infomation and schema version to history job

### LINKED ISSUE #1842: should we add tableInfo into ddl job?
I think we can add the associated tableInfo into dll job (addColumn,dropColumn,addIndex, dropIndex) before we place it into history ddl job queue.

The cause is  TiDB-Binlog need to compute the new schema base on ddl job.
I don't think it can guaranteed completely correct, so load from TiDB maybe is the best way.

@shenli @coocood @iamxy

### PR DESCRIPTION
**Pull request by zimulala**
Fix #1842

### CODE REVIEW

**Comment by zimulala** [comment]
PTAL @coocood @tiancaiamao @iamxy @hanfei1991

File: ddl/index.go [hunk 1]
```diff
@@ -303,9 +304,11 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) error {
 			return errors.Trace(err)
 		}
 
+		log.Warnf("drop index ver %v", ver)
```
Comment (hanfei1991) [inline ddl/index.go]: Useless log ?

File: ddl/bg_worker.go [hunk 2]
```diff
@@ -107,8 +107,16 @@ func (d *ddl) prepareBgJob(t *meta.Meta, ddlJob *model.Job) error {
...
 		Type:     ddlJob.Type,
-		Args:     ddlJob.Args,
 	}
+
+	argsLen := len(ddlJob.Args)
+	if argsLen > 0 {
+		job.Args = make([]interface{}, argsLen-1)
+		// ddlJob.Args[0] is the schema version that isn't necessary in background job and it will make
+		// the background job of dropping schema become more complicated to handle.
+		copy(job.Args, ddlJob.Args[1:])
```
Comment (hanfei1991) [inline ddl/bg_worker.go]: Why must we copy here instead of job.Args = ddlJob.Args[1:]?

File: ddl/foreign_key.go [hunk 3]
```diff
@@ -52,6 +52,7 @@ func (d *ddl) onCreateForeignKey(t *meta.Meta, job *model.Job) error {
 		}
 		// finish this job
 		job.State = model.JobDone
+		addFinishInfo(job, ver, nil)
```
Comment (hanfei1991) [inline ddl/foreign_key.go]: Any test for create/drop foreign key ?

File: ddl/reorg.go [hunk 4]
```diff
@@ -242,6 +242,15 @@ func (d *ddl) delKeysWithPrefix(prefix kv.Key, jobType JobType, job *model.Job,
 	return count, nil
 }
 
+// addFinishInfo adds schema version and table information that are used to binlog.
```
Comment (hanfei1991) [inline ddl/reorg.go]: used for

File: ddl/reorg.go [hunk 5]
```diff
@@ -242,6 +242,15 @@ func (d *ddl) delKeysWithPrefix(prefix kv.Key, jobType JobType, job *model.Job,
...
+func addFinishInfo(job *model.Job, ver int64, tblInfo *model.TableInfo) {
```
Comment (hanfei1991) [inline ddl/reorg.go]: Why not make this function as the model.job's member?

File: ddl/table.go [hunk 6]
```diff
@@ -131,9 +132,9 @@ func (d *ddl) onDropTable(t *meta.Meta, job *model.Job) error {
 			break
 		}
 		// finish this job
-		job.Args = []interface{}{tblInfo}
 		job.State = model.JobDone
 		job.SchemaState = model.StateNone
+		addFinishInfo(job, ver, nil)
```
Comment (hanfei1991) [inline ddl/table.go]: Why not pass tblInfo?

File: ddl/schema_test.go [hunk 7]
```diff
@@ -188,8 +242,16 @@ func (s *testSchemaSuite) TestSchemaT(c *C) {
 		updatedCount++
 	}
 	d.setHook(tc)
-	job = testDropSchema(c, ctx, d, dbInfo)
+	job, v := testDropSchema(c, ctx, d, dbInfo)
 	testCheckSchemaState(c, d, dbInfo, model.StateNone)
+	ids := make(map[int64]struct{})
+	ids[tblInfo1.ID] = struct{}{}
+	ids[tblInfo2.ID] = struct{}{}
+	checkHistoryJobArgs(c, ctx, job.ID,
```
Comment (hanfei1991) [inline ddl/schema_test.go]: unify the format for this function call.

File: ddl/schema_test.go [hunk 8]
```diff
@@ -133,7 +136,58 @@ func testCheckJobCancelled(c *C, d *ddl, job *model.Job) {
...
-func (s *testSchemaSuite) TestSchemaT(c *C) {
+func getSchemaVer(c *C, ctx context.Context) int64 {
+	txn, err := ctx.GetTxn(true)
+	c.Assert(err, IsNil)
+	c.Assert(txn, NotNil)
+	defer ctx.RollbackTxn()
+	m := meta.NewMeta(txn)
+	ver, err := m.GetSchemaVersion()
+	c.Assert(err, IsNil)
+
```
Comment (hanfei1991) [inline ddl/schema_test.go]: a useless blank?

File: ddl/table.go (hunk 6 above)
Comment (zimulala) [inline ddl/table.go]: Because only alter table operations need to add tblInfo, you can read issue #1842.

File: ddl/reorg.go (hunk 5 above)
Comment (zimulala) [inline ddl/reorg.go]: I don't want this information embed model.Job. Because job's arguments are different for each operation and here only for binlog.

**Comment by zimulala** [comment]
PTAL @hanfei1991

**Comment by tiancaiamao** [comment]
LGTM

**Comment by zimulala** [comment]
PATL @XuHuaiyu

**Comment by hanfei1991** [comment]
Will truncate table need add finish job info ?

File: ddl/schema_test.go [hunk 9]
```diff
@@ -133,7 +136,55 @@ func testCheckJobCancelled(c *C, d *ddl, job *model.Job) {
...
+	return ver
+}
+
+type historyJobArgs struct {
+	ver    int64
+	tbl    *model.TableInfo
+	tblIDs map[int64]struct{}
+}
+
+func checkHistoryJobArgs(c *C, ctx context.Context, id int64, args *historyJobArgs) error {
```
Comment (XuHuaiyu) [inline ddl/schema_test.go]: The return value seems always be nil?
Can the return value declaration of this function be removed.

**Comment by iamxy** [comment]
LGTM

**Comment by zimulala** [comment]
Add the function now. @hanfei1991

**Comment by hanfei1991** [comment]
Then add test ? @zimulala

**Comment by zimulala** [comment]
Truncate table doesn't have test in ddl package, I will update ddl test and add this test in the next PR.

**Comment by hanfei1991** [comment]
LGTM

**Comment by XuHuaiyu** [comment]
LGTM

**Pull request merged**